### PR TITLE
Make permission approval redirect flow consistent

### PR DESCRIPTION
**Pull request by rekmarks**
Previously, the permission approval redirect flow exhibited some inconsistencies. For instance, when cancelling in the fullscreen or non-notification popup UI, the redirect page would be displayed after cancelling on the second page of the connect flow, but not the first.

The redirect page is now only displayed on permission approval, for all UIs.

When cancelling, the notification UI is closed immediately. The other UIs use an immediate call to `history.push(DEFAULT_ROUTE)`. `rejectPermissionsRequest` now forces a state update, so that the UI is aware that the pending request was removed after being cancelled.

The redirect flow component was extracted from the `permission-page-container`, where it was previously rendered in a class method.

### CODE REVIEW

File: ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js [hunk 1]
```diff
@@ -5,13 +5,14 @@ import IconWithFallBack from '../../ui/icon-with-fallback'
...
     icon: PropTypes.string,
-    iconName: PropTypes.string.isRequired,
+    iconName: PropTypes.string,
     headerTitle: PropTypes.node,
     headerText: PropTypes.string,
   }
 
   static defaultProps = {
     icon: null,
+    iconName: 'Unknown External Domain',
```
Comment (rekmarks) [inline ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js:15]: There's a hot second where this prop becomes undefined because the permissions request is deleted, if the request is approved in another UI instance.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 2]
```diff
@@ -76,9 +78,38 @@ export default class PermissionConnect extends Component {
...
+    }
+
+    const environmentType = getEnvironmentType()
+    if (
+      environmentType === ENVIRONMENT_TYPE_FULLSCREEN ||
+      environmentType === ENVIRONMENT_TYPE_NOTIFICATION
+    ) {
+      window.addEventListener('beforeunload', this.beforeUnload)
+    }
+  }
```
Comment (rekmarks) [inline ui/app/pages/permissions-connect/permissions-connect.component.js:102]: This was just moved here. It's functionally identical.

File: app/_locales/en/messages.json [hunk 3]
```diff
@@ -76,9 +76,6 @@
   "cancelling": {
     "message": "Cancelling..."
   },
-  "cancelledConnectionWithMetaMask": {
-    "message": "Cancelled Connection With MetaMask"
-  },
```
Comment (rekmarks) [inline app/_locales/en/messages.json:81]: This and other deleted messages were already not used in practice.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 4]
```diff
@@ -99,62 +130,39 @@ export default class PermissionConnect extends Component {
...
+    const timeout = approved ? APPROVE_TIMEOUT : REJECT_TIMEOUT
+
     if (getEnvironmentType() === ENVIRONMENT_TYPE_NOTIFICATION) {
       setTimeout(async () => {
         global.platform.closeCurrentWindow()
-      }, 1500)
-    } else if (location.pathname === confirmPermissionPath) {
+      }, timeout)
+    } else {
       setTimeout(async () => {
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: _Note:_ `redirectFlow` has been renamed to `redirect`.

So, I just tested this again, and calling `history.push(DEFAULT_ROUTE)` (directly or via `redirect`) _caused_ the jankyness, in particular if on the second part of the connect form.

I believe this is because if it's called directly, there's still a permission request pending, and the component has to update (with the permissions request now gone), in order to be routed back to `DEFAULT_ROUTE`.

I solved this by only calling `redirectFlow` inside `componentDidUpdate`. What happens when calling `cancelPermissionsRequest` now is:

1. We dispatch the `rejectPermissionsRequest` action.
2. The permissions request is removed.
3. All `permissions-connect` receive new props, without any permissions request. This causes `redirectFlow` to be called in `componentDidUpdate`.
4. The components redirect appropriately.

They now feel responsive enough to me.

**Comment by rekmarks** [comment]
I believe all outstanding issues are address. The cancellation flow is reasonably snappy, and the notification UI handles multiple pending permissions requests.

I'm unclear why we're handling permissions requests via routes with request IDs, as opposed to e.g. transactions, where multiple pending requests are handled in the order received, and can be navigated.

That's something that we can address later, however.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 5]
```diff
@@ -53,43 +49,75 @@ export default class PermissionConnect extends Component {
...
+    getRequestAccountTabIds()
+
+    if (!permissionsRequest) {
+      return history.push(DEFAULT_ROUTE)
+    }
+
+    const environmentType = getEnvironmentType()
+    if (environmentType === ENVIRONMENT_TYPE_NOTIFICATION) {
+      window.addEventListener('beforeunload', this.beforeUnload)
+    }
```
Comment (rekmarks) [inline ui/app/pages/permissions-connect/permissions-connect.component.js:90]: `componentDidMount` was moved. The only functional change is here: only add the event listener for the notification UI.

File: ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import PropTypes from 'prop-types'
+import IconWithFallBack from '../../../components/ui/icon-with-fallback'
+import { I18nContext } from '../../../contexts/i18n'
+
+export default function PermissionsRedirect ({ domainMetadata }) {
+
+  const t = useContext(I18nContext)
+
+  return (
+    <div className="page-container permissions-redirect-container">
```
Comment (Gudahtt) [inline ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js:11]: Nit: These classes aren't the most BEM-like. `-container` sounds like it should be an element, not a block (because it isn't really a "thing" on its own), and there seems no reason to split these components between two different blocks (`permission-result` and `permission-redirect-container`).

I'm guessing some of this was pre-existing though.

File: ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js (hunk 1 above)
Comment (Gudahtt) [reply ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js:15]: Is this still a problem? We've solved this in other places by caching the relevant prop in state, so we can show it after deletion. But it'd be nice to avoid it altogether if possible.

Showing nothing might be better than this. "Unknown External Domain" isn't the most helpful thing to show, and it's also not localized. Showing nothing could be accomplished using `<Redirect to={path}>` in place of `history.push` - the `Redirect` component navigates and renders nothing.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 7]
```diff
@@ -53,43 +49,75 @@ export default class PermissionConnect extends Component {
...
+    } = this.props
+    const {
+      redirecting,
+      origin,
+      targetDomainMetadata: savedMetadata,
+    } = this.state
+
+    if (
+      permissionsRequest &&
+      savedMetadata.name !== targetDomainMetadata?.name
```
Comment (Gudahtt) [inline ui/app/pages/permissions-connect/permissions-connect.component.js:107]: Keeping state in sync with props is better handled by `getDeriveStateFromProps`. `componentDidUpdate` runs _after_ the update, so this `targetDomainMetadata` state will be delayed in updating.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 8]
```diff
@@ -99,63 +127,39 @@ export default class PermissionConnect extends Component {
...
     this.setState({
       redirecting: true,
-      permissionAccepted: accepted,
+      permissionsApproved: approved,
     })
     this.removeBeforeUnload()
 
-    if (getEnvironmentType() === ENVIRONMENT_TYPE_NOTIFICATION) {
-      setTimeout(async () => {
+    const doRedirect = () => {
```
Comment (Gudahtt) [inline ui/app/pages/permissions-connect/permissions-connect.component.js:139]: Nit: Better to move this function to module scope or make it a static method, so it's not constructed anew each time this is called.

File: ui/app/pages/permissions-connect/permissions-connect.component.js [hunk 9]
```diff
@@ -99,63 +127,39 @@ export default class PermissionConnect extends Component {
...
 
   cancelPermissionsRequest = async (requestId) => {
 
-    const { history, rejectPermissionsRequest } = this.props
+    const { rejectPermissionsRequest } = this.props
 
     if (requestId) {
       await rejectPermissionsRequest(requestId)
-
-      if (getEnvironmentType() === ENVIRONMENT_TYPE_NOTIFICATION) {
```
Comment (Gudahtt) [inline ui/app/pages/permissions-connect/permissions-connect.component.js:154]: Why were these removed?

Comment (Gudahtt) [reply ui/app/pages/permissions-connect/permissions-connect.component.js:154]: Oh I see, this is explained elsewhere

File: ui/app/pages/permissions-connect/permissions-connect.component.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: It sounds like you've just _delayed_ calling `history.push(DEFAULT_ROUTE)` by doing this. I'm not sure I understand how waiting for the next render before calling it can be faster.

[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: <details>
<summary>Redirect in "componentDidUpdate"</summary>

</details>

<details>
<summary>Redirect in "cancelPermissionsRequest"</summary>

</details>

[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: Hmm. I don't understand that second case - it's showing a flash of the list of accounts to connect to, still showing the `origin` of the permission request that should no longer exist :thinking:

[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: Yeah, I'm not totally sure why it's doing that, but that's what happens when `this.redirect(false)` is called inside `cancelPermissionsRequest` after `await rejectPermissionsRequest()`.

That's what I was getting at with this:
> I just tested this again, and calling `history.push(DEFAULT_ROUTE)` (directly or via `redirect`) _caused_ the jankyness, in particular if on the second part of the connect form.

[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: And so while you're exactly right that we'd expect things to appear less responsive with the changes I made, we in fact see the opposite.

File: ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js (hunk 6 above)
Comment (rekmarks) [reply ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js]: Yeah, a lot of it was pre-existing. Not sure if the names are an improvement or not, but the HTML and CSS have both been simplified.

File: ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js (hunk 1 above)
Comment (rekmarks) [reply ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js:15]: This was just throwing an error, it was never really visible. I changed it to an empty string and that should be fine.

File: ui/app/pages/permissions-connect/permissions-connect.component.js (hunk 8 above)
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js]: d8e7fb4

File: ui/app/pages/permissions-connect/permissions-connect.component.js (hunk 7 above)
Comment (rekmarks) [reply ui/app/pages/permissions-connect/permissions-connect.component.js:99]: d8e7fb4

File: ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js (hunk 1 above)
Comment (rekmarks) [reply ui/app/components/app/permissions-connect-header/permissions-connect-header.component.js]: d8e7fb4

File: ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js (hunk 6 above)
Comment (rekmarks) [reply ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js]: 14a5e77

**Review by Gudahtt — DISMISSED** [review]
LGTM!

I think the cancel flow could still be improved somewhat by skipping the state update, and the lifecycle logic re: redirects being simplified to only handle approval. But this is still a definite improvement

File: ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+export default function PermissionsRedirect ({ domainMetadata }) {
+
+  const t = useContext(I18nContext)
+
+  return (
+    <div className="permissions-redirect">
+      <div className="permissions-redirect__result">
+        { t('connecting') }
+        <div className="permissions-redirect__result__icons">
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/pages/permissions-connect/redirect/permissions-redirect.component.js]: Okay, _now_ it should be good! 😅

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by brad-decker — APPROVED** [review]
